dramatiq/middleware/shutdown.py

```diff
@@ -0,0 +1,69 @@
+# This file is a part of Dramatiq.
+#
+# Copyright (C) 2017,2018 CLEARTYPE SRL <bogdan@cleartype.io>
+#
+# Dramatiq is free software; you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or (at
+# your option) any later version.
+#
+# Dramatiq is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
+# License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import threading
+
+from ..logging import get_logger
+from .threading import ThreadExceptingMiddleware
+
+
+class Shutdown(BaseException):
+    """
+    Raised asynchronously inside worker threads when an actor is expected
+    to cleanup during worker shutdown.
+
+    This is *not* a subclass of DramatiqError to avoid it being caught
+    unintentionally.
+    """
+
+
+class ShutdownNotifications(ThreadExceptingMiddleware):
+    """
+    Middleware that notifies actors that the worker has been signaled
+    for termination, so that they may gracefully end their execution.
+    Currently, this is only available on CPython.
+
+    Note:
+      This class inherits `ThreadExceptingMiddleware` and its caveats.
+
+    """
+
+    def __init__(self):
+        self.logger = get_logger(__name__, type(self))
+        self.notifications = set()
+
+    @property
+    def actor_options(self):
+        return {"notify_shutdown"}
+
+    def before_worker_shutdown(self, broker, worker):
+        self.logger.debug("Sending shutdown notification to worker threads...")
+        for thread_id in self.notifications:
+            self.logger.info("Worker shutdown notification. Raising exception in worker thread %r.", thread_id)
+            self.raise_thread_exception(thread_id, Shutdown)
+
+    def before_process_message(self, broker, message):
+        actor = broker.get_actor(message.actor_name)
+
+        if message.options.get("notify_shutdown") or actor.options.get("notify_shutdown"):
+            self.notifications.add(threading.get_ident())
+
+    def after_process_message(self, broker, message, *, result=None, exception=None):
+        thread_id = threading.get_ident()
+
+        if thread_id in self.notifications:
+            self.notifications.remove(thread_id)
```

Comment:
You should add `after_skip_message = after_process_message` underneath this.

Reply:
Should this also apply to the `TimeLimit` middleware?

Reply:
Yes, it should. Looks like I missed it when I added `SkipMessage`.